modules/spring-config/src/main/java/org/mule/runtime/config/spring/MuleHierarchicalBeanDefinitionParserDelegate.java

```diff
@@ -154,6 +210,83 @@ public BeanDefinition parseCustomElement(Element element, BeanDefinition parent)
         }
     }
 
+    //TODO MULE-9638 Remove this ugly code since it's not going to be needed anymore.
+    private void setComponentModelTypeFromBeanDefinition(BeanDefinition finalChild, ComponentModel componentModel)
+    {
+        if (componentModel != null) //This condition is needed when we are parsing something unrelated to mule. See ReferenceTestCase
+        {
+            if (finalChild != null)
+            {
+                try
+                {
+                    Class<?> type = Class.forName(finalChild.getBeanClassName());
+                    if (FactoryBean.class.isAssignableFrom(type))
+                    {
+                        try
+                        {
+                            //When the FactoryBean implementation implements the FactoryBean directly.
+                            type = (Class<?>) ((ParameterizedType) Class.forName(finalChild.getBeanClassName()).getGenericInterfaces()[0]).getActualTypeArguments()[0];
+                        }
+                        catch (Exception e2)
+                        {
+                            try
+                            {
+                                //When the FactoryBean implementation extends a Class that implements FactoryBean.
+                                type = (Class<?>) ((ParameterizedType) Class.forName(finalChild.getBeanClassName()).getGenericSuperclass()).getActualTypeArguments()[0];
+                            }
+                            catch(Exception e3)
+                            {
+                                try
+                                {
+                                    //We get the type directly from a FactoryBean instance if it has a default constructor.
+                                    type = ((FactoryBean)type.newInstance()).getObjectType();
+                                }
+                                catch (InstantiationException e)
+                                {
+                                    type = Object.class;
+                                }
+                            }
+                        }
+                    }
+                    componentModel.setType(type);
+                }
+                catch (Exception e)
+                {
+                    throw new RuntimeException(e);
+                }
+            }
+        }
+    }
+
+    private boolean shouldUseNewMechanism(Element element)
```

Comment:
could you add some comments explaining which is the criteria for the new mechanism to be used?